packages/block-editor/src/components/use-on-block-drop/index.js

```diff
@@ -232,6 +232,21 @@ export default function useOnBlockDrop(
 	} = useDispatch( blockEditorStore );
 	const registry = useRegistry();
 
+	const { targetRootClientId } = useSelect(
+		( select ) => {
+			const firstParent = select( blockEditorStore ).getBlockParents(
+				originalTargetRootClientId,
+				true
+			)[ 0 ];
+			return {
+				targetRootClientId: [ 'before', 'after' ].includes( operation )
+					? firstParent
+					: originalTargetRootClientId,
+			};
+		},
+		[ originalTargetRootClientId, operation ]
+	);
```

Comment:
It might not be best performance wise to call this `useSelect` here, but by figuring out the right client id to use here, we ensure that further down `onBlockDrop` handles dealing with block indexes while factoring in blocks at the same level that share a block index.

This resolves an issue where if the block you're dragging is already just before the top area of a container block that is directly beneath it, the block doesn't suddenly get dropped _after_ that container block but instead remains in place. It's a bit of an edge case, so good if we can have the existing logic handle it.